File: 0448-find-all-numbers-disappeared-in-an-array/0448-find-all-numbers-disappeared-in-an-array.java
class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        ArrayList <Integer> result = new ArrayList<>();
        
        HashMap<Integer, Integer> map = new HashMap<>();
        
        for(int i = 0; i < nums.length; i++){
            map.put(nums[i], i); 
        }
        
        for(int i = 0; i < nums.length; i++){
            if(!map.containsKey(i + 1)){
                result.add(i + 1);
            }
        }
        return result;
    }
}

// TC : O(2n)
//SC : O(n)

File: 0448-find-all-numbers-disappeared-in-an-array/0448-find-all-numbers-disappeared-in-an-array.py
class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        result = []
        n = len(nums)
        
        for i in range(n):
            idx = abs(nums[i]) - 1
            
            if nums[idx] > 0:
                nums[idx] *= -1
        
        for i in range(n):
            if nums[i] > 0:
                result.append(i + 1)
        
        return result

# TC: O(2n)
# SC: O(1)